### ISSUE TITLE
Installation Issue

### ISSUE THREAD

**Issue by william-denault** [rstudio/keras#1412]
Hello I aam trying to install Tensorflow on a Unix server that use to do some computation. I have followed your instruction below and I get the following error message when I run ` tensorflow::as_tensor("Hello World")`

```


2024-03-04 06:52:39.913688: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2024-03-04 06:52:39.975193: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.
2024-03-04 06:52:39.975892: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-03-04 06:52:41.327400: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
Error: Valid installation of TensorFlow not found.

Python environments searched for 'tensorflow' package:
 /home/wdenault/.pyenv/versions/3.9.18/bin/python3.9

Python exception encountered:
 Traceback (most recent call last):
  File "/home/wdenault/R/x86_64-pc-linux-gnu-library/4.1/reticulate/python/rpytools/loader.py", line 119, in _find_and_load_hook
    return _run_hook(name, _hook)
  File "/home/wdenault/R/x86_64-pc-linux-gnu-library/4.1/reticulate/python/rpytools/loader.py", line 93, in _run_hook
    module = hook()
  File "/home/wdenault/R/x86_64-pc-linux-gnu-library/4.1/reticulate/python/rpytools/loader.py", line 117, in _hook
    return _find_and_load(name, import_)
  File "/home/wdenault/.virtualenvs/r-tensorflow/lib/python3.9/site-packages/tensorflow/__init__.py", line 52, in <module>
    from ._api.v2 import compat
  File "/home/wdenault/R/x86_64-pc-linux-gnu-library/4.1/reticulate/python/rpytools/loader.py", line 119, in _find_and_load_hook
    return _run_hook(name, _hook
```
Here is the output of sessionInfo()
```


sessionInfo()
R version 4.1.0 (2021-05-18)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.3.13-el7-x86_64/lib/libopenblas_haswellp-r0.3.13.so

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.12            lattice_0.20-44        here_1.0.1
 [4] png_0.1-8              rprojroot_2.0.4        grid_4.1.0
 [7] lifecycle_1.0.4        jsonlite_1.8.8         magrittr_2.0.3
[10] tfruns_1.5.2           rlang_1.1.3            cli_3.6.2
[13] whisker_0.4.1          Matrix_1.6-5           reticulate_1.35.0.9000
[16] compiler_4.1.0         base64enc_0.1-3        tensorflow_2.15.0.9000
```

Thank you in advance.

Best regards
William

Many installation issues are resolved by running the following in a **fresh R session** (you can restart R in Rstudio with Ctrl+Shift+F10) :
```R
# install the development version of packages, in case the
# issue is already fixed but not on CRAN yet.
install.packages("remotes")
remotes::install_github(sprintf("rstudio/%s", c("reticulate", "tensorflow", "keras")))
if (is.null(reticulate::virtualenv_starter()))
  reticulate::install_python()
keras3::install_keras()
```

Test to see if installation was successful.
```R
tensorflow::as_tensor("Hello World")
```

If the above snippet succeeded and you saw something like `tf.Tensor(b'Hello World', shape=(), dtype=string)`, then :tada:, you've successfully installed Tensorflow.

If the above installation failed, please gather some diagnostic info:
```R
reticulate::py_config()
tensorflow::tf_config()
reticulate::import("tensorflow")
reticulate::py_last_error()
sessionInfo()
```

Please copy and paste the FULL OUTPUT of running all three snippets, and be sure to enclose the output lines with three backticks (```) for monospace formatting.

**Comment by t-kalinowski** [comment]
Can you please post the full output from:
```r
reticulate::install_python()
keras3::install_keras()
```

**Comment by william-denault** [comment]
Hi @t-kalinowski 
Thanks for the quick feedback. here the output.
```

>tensorflow::as_tensor("Hello World")
2024-03-04 12:50:16.002903: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
Error: Valid installation of TensorFlow not found.

Python environments searched for 'tensorflow' package:
 /home/wdenault/.pyenv/versions/3.9.18/bin/python3.9

Python exception encountered:
 Traceback (most recent call last):
  File "/home/wdenault/R/x86_64-pc-linux-gnu-library/4.1/reticulate/python/rpytools/loader.py", line 119, in _find_and_load_hook
    return _run_hook(name, _hook)
  File "/home/wdenault/R/x86_64-pc-linux-gnu-library/4.1/reticulate/python/rpytools/loader.py", line 93, in _run_hook
    module = hook()
  File "/home/wdenault/R/x86_64-pc-linux-gnu-library/4.1/reticulate/python/rpytools/loader.py", line 117, in _hook
    return _find_and_load(name, import_)
  File "/home/wdenault/.virtualenvs/r-keras/lib/python3.9/site-packages/tensorflow/__init__.py", line 54, in <module>
    from tensorflow._api.v2 import compat
  File "/home/wdenault/R/x86_64-pc-linux-gnu-library/4.1/reticulate/python/rpytools/loader.py", line 119, in _find_and_load_hook
    return _run_hook(name,

```

**Comment by t-kalinowski** [comment]
Hi, Can you please post the output from running these commands:
```r
reticulate::install_python()
keras3::install_keras()
```

(Not the output from running `tensorflow::as_tensor("Hello World")`)

**Comment by t-kalinowski** [comment]
Thanks, it looks like installation succeeded.

I think the issue is: 
  - the `tensorflow` R package is looking for venv `r-tensorflow`, by default created by `tensorflow::install_tensorflow()` and `keras::install_keras()`
  - With Keras 3, `keras3::install_keras()` no longer creates venv `r-tensorflow`, instead creating a venv named `r-keras`. This is because keras is now multi backend.

You can explicitely choose a venv:
```r
library(keras3)
library(tensorflow)

use_virtualenv("r-keras")
```

**Comment by william-denault** [comment]
Ok but it seems that keras 3 has a different way to be written than keras do you have any doc about it 

For example this would work in keras but not in keras 3
```

param_nnet =keras_model_sequential() %>%
  layer_dense(units = 64,
              activation = 'relu',
              input_shape = c(1)) %>%
  layer_dense(units = 64,
              activation = 'relu' ) %>% 
  layer_dense(units = 64,
              activation = 'relu' ) %>%
  layer_dense(units = 64,
              activation = 'relu' ) %>%
  layer_dense(units = 10,
              activation = 'softmax')
```

with keras 3 I got the following output

```
Error in do.call(LayerClass, args) :
  'what' must be a function or character string
```
which is not really indicative of what is going. Do you have any suggestion?

**Comment by t-kalinowski** [comment]
The ssl error is likely due to the fact that you're on a relatively old OS (CentOS 7), but trying to use a relatively new Python. I think usually `pyenv` (what powers  `reticulate::install_python()`)takes care of issues like this, or reports useful error messages for how to resolve. A good first step would be to run 
```r
reticulate::install_python(force = TRUE)
```
And read the output carefully. See if there is anything that looks like an error message related to SSL. 

This link might also be helpful for troubleshooting. https://github.com/pyenv/pyenv/wiki/Common-build-problems#error-the-python-ssl-extension-was-not-compiled-missing-the-openssl-lib

Bottom line, you don't have a valid Python installation that's currently not EOL'd.
